### ISSUE TITLE
HTTP Lifecycle hook tests flake on multinode clusters

### ISSUE THREAD

**Issue by claudiubelu** [kubernetes/kubernetes#101062]
#### Which jobs are flaking:

k8s-e2e-ltsc2019-containerd-flannel-sdnoverlay-master

#### Which test(s) are flaking:

[k8s.io] Container Lifecycle Hook when create a pod with lifecycle hook should execute prestart http hook properly [NodeConformance] [Conformance]
[k8s.io] Container Lifecycle Hook when create a pod with lifecycle hook should execute poststart http hook properly [NodeConformance] [Conformance]

#### Testgrid link:

https://testgrid.k8s.io/sig-windows-networking#sac1909-containerd-flannel-sdnoverlay-stable

Note: the tests have been excluded until they are fixed.

#### Reason for failure:

In the case of multinode clusters, the http server pod and the test cluster can spawn on different nodes, which can be problematic for  poststart / prestop hooks, as they are executed by the kubelet itself, and the cross-node lifecycle hook might
fail (according to the [Kubernetes network model], it is not mandatory for kubelet to be able to access pods on a different node).

However, that is not always the case, different Networking setups allow cross-node communication between kubelet and pods, which means that the above tests are passing properly (for example: https://testgrid.k8s.io/sig-windows-networking#sac1909-containerd-flannel-sdnbridge-stable), which begs the question if this behaviour should be universal or not, especially since the tests are marked as ``Conformance``, which would imply that it should work in all cases.

[Kubernetes network model]: https://kubernetes.io/docs/concepts/cluster-administration/networking/#the-kubernetes-network-model

#### Anything else we need to know:
- links to go.k8s.io/triage appreciated
- links to specific failures in spyglass appreciated

Slack: https://kubernetes.slack.com/archives/C09QYUH5W/p1617965793211800

/sig testing
/sig networking
/sig windows

**Comment by aojea** [comment]
/sig network
/sig node
/sig conformance
/assign @thockin @johnbelamaric 
/priority critical-urgent
/cc @dims @liggitt 

This is important since affects conformance and defines the networking behaviour for the poststart/ prestop hooks

**Comment by dims** [comment]
@claudiubelu i'd personally let @kubernetes/sig-network-api-reviews folks have first dibs on what we need to do here. Though from the name of the test itself, it seems like the test is NOT a sig-network test :)

**Comment by claudiubelu** [comment]
> 
> 
> @claudiubelu i'd personally let @kubernetes/sig-network-api-reviews folks have first dibs on what we need to do here. Though from the name of the test itself, it seems like the test is NOT a sig-network test :)

Indeed, it does not have any ``[sig-networking]`` tag in its name, but the tests rely on networking and on how CNIs work. As I've stated above, it's not a problem for flannel sdnbridge, but it is for flannel sdnoverlay; other CNIs might have a similar issue too.

**Comment by aojea** [comment]
> Indeed, it does not have any [sig-networking] tag in its name, but the tests rely on networking
because the network is not a thing :)  let´s tag the sig owner then
/sig node
I think that the key here is to decide if kubelet should be able to access pod in other nodes, and based on the link Claudiu added is part of the Kubernetes networking model.
I saw users that doesn't want kubelet to access even the pods on same host, so I think that is legit to say kubelet MAY reach pods in other nodes but not MUST ... but I'm not familiar with the `lifecycle hook` to fully understand the context.
If kubelet "may" reach pods in the other nodes, the test must only test lifecycle hooks for pods in the same node as the kubelet under test, right?

**Comment by thockin** [comment]
> according to the Kubernetes network model, it is not mandatory for kubelet to be able to access pods on a different node

This is correct.  The model also says:

"For those platforms that support Pods running in the host network (e.g. Linux): pods in the host network of a node can communicate with all pods on all nodes without NAT"

I believe that Windows host network is not supported, right?

**Comment by claudiubelu** [comment]
> 
> 
> > according to the Kubernetes network model, it is not mandatory for kubelet to be able to access pods on a different node
> 
> This is correct. The model also says:
> 
> "For those platforms that support Pods running in the host network (e.g. Linux): pods in the host network of a node can communicate with all pods on all nodes without NAT"
> 
> I believe that Windows host network is not supported, right?

Right, and there are specific tests that checks that exact scenario you've mentioned:

```
[sig-network] Networking Granular Checks: Pods should function for node-pod communication: http [LinuxOnly] [NodeConformance] [Conformance]
[sig-network] Networking Granular Checks: Pods should function for node-pod communication: udp [LinuxOnly] [NodeConformance] [Conformance]
```

However, the flaky tests I've mentioned in the issue require kubelet to be able to dial-in into a webserver pod for a different pod's poststart / prestop HTTP lifecycle hooks. There are couple of things regarding this:

1. First of all, in order to ensure kubelet would be able to run the HTTP lifecycle hooks, we'd have to make sure both pods are on the same node, as that is a supported scenario, according to the Kubernetes Network Model. I have sent a PR for this above. Alternatively, a Cluster IP could probably work as well, but IMO, ensuring that both pods end up on the same node is simpler and easier to clean up.
2. Secondly, it's interesting that those HTTP lifecycle hook tests work on different networking setups, *without* any pods having HostNetwork enabled, even though you've mentioned they should (e.g.: Linux / Windows with flannel l2bridge, and even Linux with flannel overlay). Now, my question is: is this behaviour a feature, or a bug, especially since the tests are marked as ``[Conformance]``?

**Comment by aojea** [comment]
> "For those platforms that support Pods running in the host network (e.g. Linux): pods in the host network of a node can communicate with all pods on all nodes without NAT"

oh, I missed that part, so in Linux kubelet must be able to communicate with pods on all nodes (becuase it is running in the hostnetwork)

If that is correct, it seems that one solution is to keep the test for Linux  and tag it as LinuxOnly, and a new test for WindowsOnly that only spawn the pods under test in the same node

**Comment by thockin** [comment]
For the purposes of testing, I don't mind forcing the pods onto the same node., but let's document the heck out of that at the code-site?

**Comment by jsturtevant** [comment]
> I believe that Windows host network is not supported, right?

correct but it will be supported once HostProcess containers are merged in 1.22: https://github.com/kubernetes/kubernetes/pull/99576

> If that is correct, it seems that one solution is to keep the test for Linux and tag it as LinuxOnly, and a new test for WindowsOnly that only spawn the pods under test in the same node

Only in Overlay networks (ie Flannel where these tests are flaking) does the pod have to be on the same node.  It doesn't require host network mode and these tests are [not flaky on l2bridge setup](https://testgrid.k8s.io/sig-windows-signal#aks-engine-windows-containerd-1.21&include-filter-by-regex=pod.with.lifecycle.hook%20).  I would be more inclined to disable this test for Windows Overlay test setups and have a sig-windows test for the overlay functionality.

I would also question if the lifecycle hooks only running on the same node is useful and if it should just be documented that it doesn't work in the Windows Overlay setup.

**Comment by SergeyKanzhelev** [comment]
@fromanirh is looking at test PRs from SIG Node. I will let sig-net triage this issue

**Comment by thockin** [comment]
Next steps on this?

**Comment by claudiubelu** [comment]
> 
> 
> Next steps on this?

We still haven't merged this: https://github.com/kubernetes/kubernetes/pull/101063

I also had this, but it has been decided to be blocked: https://github.com/kubernetes/kubernetes/pull/101060

**Comment by aojea** [comment]
> We still haven't merged this: #101063

requires conformance approval, at this point I think that has to be postponed to 1.23

> I also had this, but it has been decided to be blocked: #101060

right, since that behavior is not required, we shouldn't test it, because is legit to fail and having a test for something that can fail can create confusion :smile:
